PFR/Homomorphism.lean

```diff
@@ -94,8 +96,168 @@ lemma goursat (H : AddSubgroup (G × G')): ∃ (H₀ : AddSubgroup G) (H₁ : Ad
   exact h_bij'
   rw [Nat.card_eq_of_bijective bij h_bij, Nat.card_prod H₀ H₁]
 
+
+/- TODO: Find an appropriate home for these lemmas -/
+lemma Nat.card_image_le {α β: Type*} {s : Set α} {f : α → β} (hs : s.Finite) :
+    Nat.card (f '' s) ≤ Nat.card s := by
+  simp only [Set.Nat.card_coe_set_eq]
+  exact Set.ncard_image_le hs
+
+lemma Nat.card_singleton_prod {α β : Type*} (a : α) (B : Set β) : Nat.card ({a} ×ˢ B) = Nat.card B := by
+  by_cases hB : Set.Finite B
+  · rw[Set.singleton_prod, Nat.card_image_of_injective (Prod.mk.inj_left a) hB]
+  · rw[Set.Infinite.card_eq_zero hB, Set.Infinite.card_eq_zero <| Set.Infinite.prod_right hB ⟨a,by rfl⟩]
+
+lemma Nat.card_prod_singleton {α β : Type*} (A : Set α) (b : β) : Nat.card (A ×ˢ {b}) = Nat.card A := by
+  by_cases hA : Set.Finite A
+  · rw[Set.prod_singleton, Nat.card_image_of_injective (Prod.mk.inj_right b) hA]
+  · rw[Set.Infinite.card_eq_zero hA, Set.Infinite.card_eq_zero <| Set.Infinite.prod_left hA ⟨b,by rfl⟩]
+
+@[to_additive]
+lemma Nat.card_inv [Group G] (S : Set G) : Nat.card (S⁻¹ : Set G) = Nat.card S := by
+  rw [←Set.image_inv]
+  apply Nat.card_image_of_injective
+  · exact inv_injective
+  · exact Set.toFinite S
+
+open Set Fintype in
+-- variable [DecidableEq G] [DecidableEq G'] in
 /-- Let $f: G \to G'$ be a function, and let $S$ denote the set
 $$ S := \{ f(x+y)-f(x)-f(y): x,y \in G \}.$$
 Then there exists a homomorphism $\phi: G \to G'$ such that
 $$ |\{ f(x) - \phi(x)\}| \leq 4 |S|^{24}.$$ -/
-theorem homomorphism_pfr (f : G → G') (S : Set G') (hS: ∀ x y : G, f (x+y) - (f x) - (f y) ∈ S): ∃ (φ : G →+ G') (T : Set G'), Nat.card T ≤ 4 * (Nat.card S)^24 ∧ ∀ x : G, (f x) - (φ x) ∈ T := by sorry
+theorem homomorphism_pfr (f : G → G') (S : Set G') (hS: ∀ x y : G, f (x+y) - (f x) - (f y) ∈ S):
+  ∃ (φ : G →+ G') (T : Set G'), Nat.card T ≤ 4 * (Nat.card S:ℝ)^(24:ℝ) ∧ ∀ x : G, (f x) - (φ x) ∈ T := by
+  classical
+  let A := graph f
+
+  let B := A + ({0}×ˢ(-S))
```

Comment:
Does it help to write it like this?
```suggestion
  let B := A - {0} ×ˢ S
```

Reply:
I don't think it matters much, I think I avoided this because `Nat.card_sub_le` is missing. It's probably worth adding anyway, though.